presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java

```diff
@@ -224,17 +231,41 @@ public long getValidationCpuNanos()
 
     private ListenableFuture<Collection<Slice>> doFinish()
     {
-        ImmutableList.Builder<Slice> partitionUpdates = ImmutableList.builder();
+        ImmutableList.Builder<PartitionUpdate> partitionUpdatesBuilder = ImmutableList.builder();
         List<Callable<Object>> verificationTasks = new ArrayList<>();
         for (HiveWriter writer : writers) {
             writer.commit();
-            PartitionUpdate partitionUpdate = writer.getPartitionUpdate();
-            partitionUpdates.add(wrappedBuffer(partitionUpdateCodec.toJsonBytes(partitionUpdate)));
+            partitionUpdatesBuilder.add(writer.getPartitionUpdate());
             writer.getVerificationTask()
                     .map(Executors::callable)
                     .ifPresent(verificationTasks::add);
         }
-        List<Slice> result = partitionUpdates.build();
+
+        List<PartitionUpdate> partitionUpdates = partitionUpdatesBuilder.build();
+        boolean optimizedPartitionUpdateSerializationEnabled = isOptimizedPartitionUpdateSerializationEnabled(session);
+        if (optimizedPartitionUpdateSerializationEnabled) {
+            // Merge multiple partition updates for a single partition into one.
+            // Multiple partition updates for a single partition are produces when writing into a bucketed table.
```

Comment:
Yeah. Currently partition updates for the same partition are merged on coordinator: https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java#L1640. I was a little surprised to see that the PartitioUpdate's are not merged on the worker first before sending to the coordinator (or Driver in case of Presto on Spark)